File: bus-route/shaun-gy-bfs.java
//shaun java code from c++
//idea check c++ code
//runtime 45ms, faster than 64%, mem less than 76%
// java Queue is abstract, = new LinksedList<>()
// push is offer, front is peek, pop is poll, empty --isEmpty
// vector is List, push_back  ---- add
// unordered_set  --- HashSet,  insert --- add
class Solution {
    public int numBusesToDestination(int[][] routes, int S, int T) {
        if (S == T) return 0; //no need to get on bus
        
        Map<Integer,List<Integer>> stop2bus = new HashMap<>();
        for (int bus = 0; bus < routes.length; ++bus) {
            for (int stop: routes[bus]) {
                stop2bus.putIfAbsent(stop, new ArrayList<Integer>());
                stop2bus.get(stop).add(bus);
            }
        }
        
        //bfs level by level from S to T
        Queue<Integer> q = new LinkedList<>();
        Set<Integer> visited = new HashSet<>();
        q.offer(S);
        int buses = 0;
        while (!q.isEmpty()) {
            buses++;
            int size = q.size();
            for (int i = 0; i < size; ++i) {
                int stop = q.peek(); q.poll();
                
                for (int bus: stop2bus.get(stop)) {
                    if (visited.contains(bus)) continue;
                    visited.add(bus);
                    for (int next: routes[bus]) {
                        if (next == T) return buses;
                        
                        //not visited
                        q.offer(next);
                        
                    }
                    
                }
            }
            
            
        }
        //not yet reach T, not possible, return -1
        return -1;
    }
}

File: bus-route/shaun-gy-bfs.cpp
//shaun first try, wrong
//check with gy, not straightforward apply of the bfs template
//runtime 1636ms, faster than 5%, mem less than 100%
// stop2bus use int to vector<int> TLE
//          use int to unordered_set<int>, 5%
class Solution {
public:
    int numBusesToDestination(vector<vector<int>>& routes, int S, int T) {
        
        unordered_map<int,unordered_set<int>> stop2bus;
        //unordered_map<int,vector<int>> stop2bus;
        for (int bus = 0; bus < routes.size(); ++bus) {
            for (int stop: routes[bus]) {
                stop2bus[stop].insert(bus);
                //stop2bus[stop].push_back(bus);
            }
        }
        
        if (S == T) return 0; //no need to get on bus
        
        //bfs level by level from S to T
        queue<int> q;
        unordered_set<int> visited;
        int level = 0;
        q.push(S);
        visited.insert(S);
        while (!q.empty()) {
            level++;
            int size = q.size();
            for (int i = 0; i < size; ++i) {
                int stop = q.front(); q.pop();
                
                for (int bus: stop2bus[stop]) {
                    for (int next: routes[bus]) {
                        if (next == T) return level;
                        if (visited.count(next)) continue;
                        //not visited
                        q.push(next);
                        visited.insert(next);
                    }
                    
                }
            }
            
            
        }
        //not yet reach T, not possible, return -1
        return -1;
    }
};
